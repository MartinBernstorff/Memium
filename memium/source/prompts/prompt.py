from collections.abc import Sequence
from typing import Protocol, runtime_checkable


@runtime_checkable
class BasePrompt(Protocol):
    @property
    def scheduling_uid(self) -> int:
        """UID used when scheduling the prompt. If this UID changes, the old prompt is deleted and a new prompt is created."""
        ...

    @property
    def update_uid(self) -> int:
        """The UID used to determine whether the prompt state has changed. If this UID changes, the prompt is updated on the remote."""
        ...

    @property
    def tags(self) -> Sequence[str]: ...

    @property
    def edit_url(self) -> str | None:
        """A link which, when pressed, opens a location where the prompt can be edited. E.g. opens a given Markdown document in the correct editor, or a web interface for a CMS."""
        ...


from dataclasses import dataclass


@dataclass(frozen=True)
class DestinationPrompt:
    prompt: BasePrompt
    destination_id: str
