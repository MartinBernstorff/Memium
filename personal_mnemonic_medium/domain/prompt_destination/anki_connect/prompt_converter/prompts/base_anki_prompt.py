from abc import ABC, abstractmethod
from collections.abc import Sequence
from dataclasses import dataclass

import genanki


@dataclass(frozen=True)
class AnkiCard(ABC):
    base_deck: str
    tags: Sequence[str]

    @property
    @abstractmethod
    def uuid(self) -> int:
        ...

    @property
    @abstractmethod
    def genanki_model(self) -> genanki.Model:
        ...

    @abstractmethod
    def to_genanki_note(self) -> genanki.Note:
        ...

    @property
    def deck(self) -> str:
        deck_prefix = "anki/deck/"
        deck_in_tags = (
            tag.replace(deck_prefix, "").replace("/", "::")
            for tag in self.tags
            if tag.startswith(deck_prefix)
        )
        subdeck = next(deck_in_tags, None)

        if subdeck is None:
            return self.base_deck

        return f"{self.base_deck}::{subdeck}"
