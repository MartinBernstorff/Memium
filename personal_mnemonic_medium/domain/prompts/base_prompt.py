from collections.abc import Sequence
from typing import Protocol


class BasePrompt(Protocol):
    @property
    def scheduling_uid(self) -> int:
        """UID used when scheduling the prompt. If this UID changes, the scheduling of the prompt is reset."""
        ...

    @property
    def update_uid(self) -> int:
        """The UID used to determine whether the prompt state has changed. If this UID changes, the prompt is updated on the remote."""
        ...

    @property
    def tags(self) -> Sequence[str]:
        ...


from dataclasses import dataclass


@dataclass(frozen=True)
class DestinationPrompt:
    prompt: BasePrompt
    destination_id: str
