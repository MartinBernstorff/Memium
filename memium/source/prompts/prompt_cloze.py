from collections.abc import Sequence
from dataclasses import dataclass

from ...utils.hash_cleaned_str import clean_str, hash_cleaned_str, int_hash_str
from ..document import Document
from .prompt import BasePrompt


@dataclass(frozen=True)
class ClozePrompt(BasePrompt):
    text: str

    @property
    def scheduling_uid(self) -> int:
        return hash_cleaned_str(self.text)

    @property
    def update_uid(self) -> int:
        return int_hash_str(f"{clean_str(self.text)}{self.tags}")

    @property
    def tags(self) -> Sequence[str]:
        return ()


@dataclass(frozen=True)
class ClozeWithoutDoc(ClozePrompt):
    add_tags: Sequence[str]

    @property
    def tags(self) -> Sequence[str]:
        return self.add_tags


@dataclass(frozen=True)
class ClozeFromDoc(ClozePrompt):
    text: str
    source_doc: Document

    @property
    def scheduling_uid(self) -> int:
        return int_hash_str(self.text)

    @property
    def tags(self) -> Sequence[str]:
        return self.source_doc.tags
