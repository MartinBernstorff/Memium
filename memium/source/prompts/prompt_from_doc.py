from abc import ABC
from collections.abc import Sequence
from dataclasses import dataclass
from urllib.parse import quote

from ..document import Document
from .prompt import BasePrompt


@dataclass(frozen=True)
class PromptFromDocMixin(BasePrompt, ABC):
    parent_doc: Document
    line_nr: int

    @property
    def scheduling_uid(self) -> int: ...

    @property
    def update_uid(self) -> int: ...

    @property
    def tags(self) -> Sequence[str]: ...

    @property
    def edit_url(self) -> str | None:
        return obsidian_url(self.parent_doc.title, self.line_nr)


def obsidian_url(file_title: str, line_nr: int | None) -> str:
    url = f"obsidian://advanced-uri?filename={quote(file_title)}"
    if line_nr:
        url += f"&line={line_nr}"
    return url
