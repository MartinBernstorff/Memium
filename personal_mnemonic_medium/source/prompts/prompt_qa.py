from collections.abc import Sequence
from dataclasses import dataclass

from ...utils.hash_cleaned_str import hash_cleaned_str
from ..document import Document
from .prompt import BasePrompt


@dataclass(frozen=True)
class QAPrompt(BasePrompt):
    question: str
    answer: str

    @property
    def scheduling_uid(self) -> int:
        return hash_cleaned_str(f"{self.question}_{self.answer}")

    @property
    def update_uid(self) -> int:
        return hash_cleaned_str(f"{self.question}_{self.answer}_{self.tags}")

    @property
    def tags(self) -> Sequence[str]:
        return ()


@dataclass(frozen=True)
class QAWithoutDoc(QAPrompt):
    add_tags: Sequence[str]

    @property
    def tags(self) -> Sequence[str]:
        return self.add_tags


@dataclass(frozen=True)
class QAFromDoc(QAPrompt):
    parent_doc: Document
    line_nr: int

    def __repr__(self) -> str:
        return f"{self.parent_doc.source_path}:{self.line_nr}: \n\tQ. {self.question}\n\tA. {self.answer}"

    @property
    def tags(self) -> Sequence[str]:
        return self.parent_doc.tags
