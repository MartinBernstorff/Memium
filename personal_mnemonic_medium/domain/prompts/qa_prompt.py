from collections.abc import Sequence
from dataclasses import dataclass

from ..prompt_source.document_ingesters.document import Document
from ..utils.hash_cleaned_str import hash_cleaned_str
from .base_prompt import BasePrompt


@dataclass(frozen=True)
class QAPrompt(BasePrompt):
    question: str
    answer: str

    @property
    def uid(self) -> int:
        return hash_cleaned_str(f"{self.question}_{self.answer}")

    @property
    def tags(self) -> Sequence[str]:
        return ()


@dataclass(frozen=True)
class QAWithoutDoc(QAPrompt):
    add_tags: Sequence[str]

    @property
    def tags(self) -> Sequence[str]:
        return self.add_tags


@dataclass(frozen=True)
class QAFromDoc(QAPrompt):
    parent_doc: Document
    line_nr: int

    def __repr__(self) -> str:
        return f"{self.parent_doc.source_path}:{self.line_nr}: \n\tQ. {self.question}\n\tA. {self.answer}"

    @property
    def tags(self) -> Sequence[str]:
        return self.parent_doc.tags
