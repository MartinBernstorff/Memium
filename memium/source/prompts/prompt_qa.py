from collections.abc import Sequence
from dataclasses import dataclass

from ...utils.hash_cleaned_str import clean_str, hash_str_to_int
from .prompt import BasePrompt
from .prompt_from_doc import PromptFromDocMixin


@dataclass(frozen=True)
class QAPrompt(BasePrompt):
    question: str
    answer: str

    @property
    def scheduling_uid_str(self) -> str:
        """Str used for generating the update_uid. Super helpful for debugging."""
        return f"{clean_str(self.question)}_{clean_str(self.answer)}"

    @property
    def scheduling_uid(self) -> int:
        return hash_str_to_int(self.scheduling_uid_str)

    @property
    def update_uid_str(self) -> str:
        """Str used for generating the update_uid. Super helpful for debugging."""
        return f"{self.question}_{self.answer}_{self.tags}"

    @property
    def update_uid(self) -> int:
        return hash_str_to_int(self.update_uid_str)

    @property
    def tags(self) -> Sequence[str]:
        return ()


@dataclass(frozen=True)
class QAWithoutDoc(QAPrompt):
    add_tags: Sequence[str]

    @property
    def tags(self) -> Sequence[str]:
        return self.add_tags

    @property
    def edit_url(self) -> str | None:
        return None


@dataclass(frozen=True)
class QAFromDoc(QAPrompt, PromptFromDocMixin):
    @property
    def tags(self) -> Sequence[str]:
        return self.parent_doc.tags
