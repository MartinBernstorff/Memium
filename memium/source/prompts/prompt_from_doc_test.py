from memium.source.prompts.prompt_from_doc import obsidian_url


def test_file_title_to_uri(snapshot: str):
    title = "Heap Properties"
    assert snapshot == obsidian_url(title, 5)
