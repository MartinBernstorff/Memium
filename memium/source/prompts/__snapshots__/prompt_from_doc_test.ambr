# serializer version: 1
# name: test_file_title_to_uri
  'obsidian://advanced-uri?filename=Heap%20Properties&line=5'
# ---
